### PR TITLE
System.Net.WebSockets.Client - remove race condition from ReceiveAsync cancellation test.

### PR DESCRIPTION
**Pull request by Clockwork-Muse**

### CODE REVIEW

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs [hunk 1]
```diff
@@ -125,19 +124,14 @@ public class CancelTest : ClientWebSocketTestBase
...
 
-                try
-                {
-                    await cws.ReceiveAsync(segment, cts.Token);
-                    Assert.True(false, "Receive should not complete.");
-                }
-                catch (OperationCanceledException) { }
-                catch (ObjectDisposedException) { }
-                catch (WebSocketException) { }
+                Exception e = await Record.ExceptionAsync(() => cws.ReceiveAsync(segment, cts.Token));
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: Why not use Assert.ThrowsAnyAsync?

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 2]
```diff
@@ -157,45 +157,42 @@ public class SendReceiveTest : ClientWebSocketTestBase
...
                 }
-                catch (AggregateException ag)
+                else
                 {
-                    foreach (var ex in ag.InnerExceptions)
+                    AggregateException ae = Assert.IsType<AggregateException>(e);
+
+                    Assert.All(ae.InnerExceptions, ex =>
                     {
+                        Assert.True(ex is InvalidOperationException || ex is WebSocketException,
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Why is this better than the Assert.True(false)? In this case it's simpler and less repetition.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 3]
```diff
@@ -157,45 +157,42 @@ public class SendReceiveTest : ClientWebSocketTestBase
...
                 {
-                    for (int i = 0; i < tasks.Length; i++)
-                    {
-                        tasks[i] = cws.ReceiveAsync(recvSegment, cts.Token);
-                    }
+                    tasks[i] = cws.ReceiveAsync(recvSegment, cts.Token);
+                }
 
-                    Task.WaitAll(tasks);
+                Exception e = Record.Exception(() => Task.WaitAll(tasks));
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Similarly, why not Assert.ThrowsAnyAsync? It's not only what we use elsewhere in the tests, it avoids the need for the subsequent null check.

Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Because it's supposed to be more along the lines of [AAA](https://msdn.microsoft.com/en-us/library/hh694602.aspx#Anchor_3), which ends up separating the throw/catch from examining the value (here, the exception) itself.  
I only actually did this because I'm not the first to put this in, although I wasn't particularly planning on general replacement of the existing asserts otherwise (absent a more specific directive in one way or the other).  The null check could be potentially removed, as `IsType` should catch that as well...

Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Note, also, that `Assert.Throws` _can't_ be used here, since there is no guarantee that an exception will actually be thrown (presumably based on other parameters, although I didn't look terribly far into these tests at the moment).

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 2 above)
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: I prefer doing this, because it directly states the condition we're actually testing (that the exception is one of those types), rather than having it be inferred from the leftover conditions.  Something equivalent to an `Assert.AnyOf(T actual, IEnumerable<T> expected)`, although that's probably wrong from a testing point of view ("should know exactly what the result is").

Ideally, tests should potentially be rewritten to separate the situations under which the two exceptions are thrown... however, the code being tested involves threading/cancellation, and changing it this way would involves additional work.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: > Note, also, that Assert.Throws can't be used here, since there is no guarantee that an exception will actually be thrown

That's true here (I was basing my comment on the title of this PR, saying that you were replacing Assert.True(false)s, and didn't look to see whether this case had one).  It's not in the previous case I highlighted, where we should use Assert.ThrowsAnyAsync.

And here since Assert.True(false) isn't being used, I'm not seeing what the transformation to use Record.ExceptionAsync really helps with.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: > I prefer doing this

Everyone has their own personal preferences.  If one is obviously better, than a change makes sense.  Otherwise, as in this case, I'm inclined to just keep things the way they are.  Someone else could come along and think "I prefer this other way", and change it back.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: :smile:

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs [hunk 4]
```diff
@@ -130,14 +129,9 @@ public class CancelTest : ClientWebSocketTestBase
...
-                {
-                    await cws.ReceiveAsync(segment, cts.Token);
-                    Assert.True(false, "Receive should not complete.");
-                }
-                catch (OperationCanceledException) { }
-                catch (ObjectDisposedException) { }
-                catch (WebSocketException) { }
+                Exception e = await Assert.ThrowsAnyAsync<Exception>(() => cws.ReceiveAsync(segment, cts.Token));
+                Assert.True(e is OperationCanceledException || e is ObjectDisposedException || e is WebSocketException,
+                    $"Unexpected exception type {e.GetType()}");
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: If this does fail, we end up losing data about the exception, e.g. its stack trace.  Makes me wonder if, while not ideal, the original form that was using Assert.True is better for this case, since there at least the exception is allowed to propagate if it doesn't match the criteria / catch blocks.  At the very least if we do want to keep this as an Assert.True, the message should be changed to something like ```$"Unexpected exception: {e}"``` rather than limiting it to its type.

**Comment by stephentoub** [comment]
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please

cc: @davidsh, @cipop

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs (hunk 4 above)
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: Hmm, yeah, I can see that.
Actually, after poking around a little more, I think we can just expect `WebSocketException`:
 - `ObjectDisposedException` in this context is probably a holdover from the previous issues with tokens.
 - `OperationCanceledException` is only thrown if the token is canceled before it starts `await`ing (during setup) - it's not thrown during `Receive` otherwise (from the look of it, only during `Send`).

An OCE would thus be a sign that the test didn't get into a necessary state to begin with, and should likely be re-run.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 2 above)
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: It turns out the test here is extremely misleading, and can be narrowed significantly, allowing us to eliminate this check anyways.

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs [hunk 5]
```diff
@@ -125,19 +124,12 @@ public class CancelTest : ClientWebSocketTestBase
...
-                try
-                {
-                    await cws.ReceiveAsync(segment, cts.Token);
-                    Assert.True(false, "Receive should not complete.");
-                }
-                catch (OperationCanceledException) { }
-                catch (ObjectDisposedException) { }
-                catch (WebSocketException) { }
+                var cts = new CancellationTokenSource(100);
+                // If this fails with OperationCancelledException, was not yet being awaited - rerun test.
```
Comment (davidsh) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: grammar/spelling:
OperationCancelledException ->OperationCanceledException
"was not yet being await" -> "it was not yet being awaited"

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs [hunk 6]
```diff
@@ -147,4 +139,4 @@ public class CancelTest : ClientWebSocketTestBase
             }
         }
     }
-}
+}
```
Comment (davidsh) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: nit: missing newline at end of file.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 7]
```diff
@@ -144,59 +132,36 @@ public class SendReceiveTest : ClientWebSocketTestBase
...
-                        tasks[i] = cws.ReceiveAsync(recvSegment, cts.Token);
-                    }
-
-                    Task.WaitAll(tasks);
-                    Assert.Equal(WebSocketState.Open, cws.State);
+                    tasks[i] = cws.ReceiveAsync(recvSegment, CancellationToken.None);
                 }
-                catch (AggregateException ag)
+
+                // Even though inidividual tasks complete asynchronously, they were submitted synchronously.
```
Comment (davidsh) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: spelling: "inidividual" -> "individual"

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 8]
```diff
@@ -303,4 +268,4 @@ public class SendReceiveTest : ClientWebSocketTestBase
             }
         }
     }
-}
+}
```
Comment (davidsh) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: nit: missing newline at end of file.

**Comment by stephentoub** [comment]
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please

**Comment by Clockwork-Muse** [comment]
nits addressed and pushed, but it looks like there's different behavior between Linux and Windows for the send/receive tests I modified.  Thoughts?  Relax the tests back to what they were (which hid this behavior), special case it, or investigate more and potentially conform the behavior?

**Comment by stephentoub** [comment]
>  but it looks like there's different behavior between Linux and Windows for the send/receive tests I modified

IIRC, the Windows behavior has race conditions where it can result in either exception.  I could be wrong, but I would not be surprised if running the tests a bunch on Windows with your changes would result in them failing randomly.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 9]
```diff
@@ -86,55 +85,44 @@ public class SendReceiveTest : ClientWebSocketTestBase
         {
             using (ClientWebSocket cws = await WebSocketHelper.GetConnectedWebSocket(server, TimeOutMilliseconds, _output))
             {
-                var cts = new CancellationTokenSource(TimeOutMilliseconds);
+                var tasks = new Task[2];
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Why did you change this from 10 to 2?  Part of the goal of this test is to verify behavior when there are a bunch of outstanding send operations.  Having just one outstanding significantly diminishes the value of what's being tested.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 10]
```diff
@@ -86,55 +85,44 @@ public class SendReceiveTest : ClientWebSocketTestBase
...
 
-                Task[] tasks = new Task[10];
+                // Need large buffer, or Send completes too quickly and no exception is thrown.
+                var sendBuffer = new byte[128 * 1024];
+                var segment = new ArraySegment<byte>(sendBuffer);
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: The buffer will be all zeros, yet it's being sent as Text.  It'd be better to have it filled will actual UTF8 text, as it was previously.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 11]
```diff
@@ -86,55 +85,44 @@ public class SendReceiveTest : ClientWebSocketTestBase
...
-                                (errCode == WebSocketError.InvalidState) || (errCode == WebSocketError.Success),
-                                "WebSocketErrorCode");
-                        }
-                        else
-                        {
-                            Assert.IsAssignableFrom<OperationCanceledException>(ex);
-                        }
-                    }
-                }
+                    Assert.Equal(TaskStatus.Faulted, task.Status);
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: This does not look like a safe change.  Due to timing and race conditions and the like, it's possible that any number of the operations could actually complete successfully.  Seems like this is going to be very prone to sporadic failures in the lab.

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs [hunk 12]
```diff
@@ -125,19 +124,12 @@ public class CancelTest : ClientWebSocketTestBase
...
-                try
-                {
-                    await cws.ReceiveAsync(segment, cts.Token);
-                    Assert.True(false, "Receive should not complete.");
-                }
-                catch (OperationCanceledException) { }
-                catch (ObjectDisposedException) { }
-                catch (WebSocketException) { }
+                var cts = new CancellationTokenSource(100);
+                // If this fails with OperationCancelledException, it was not yet being awaited - rerun test.
```
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: Then I need to rephrase the comment.
Going by the method name (and internals), what the method is _attempting_ to test is that cancellation that occurs once `await`ing has started throws a `WebSocketException`.  It's trying for a completely different code path.  So if we get an OCE, we didn't hit the target path, and the test is inconclusive (hence, the re-run instruction).  100 ms is an eternity to a machine - I get `WebSocketException` even if I set it to **1** ms.
Perhaps something like:
> If cancellation occurs after ReceiveAsync returns the awaitable task, it should throw WebSocketException, and not OperationCanceledException.  If this assert fails, it's possibly because of simple timing issues; try re-running the test.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 10 above)
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Okay, I can throw "random" data in there.

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs (hunk 12 above)
Comment (CIPop) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: > 100 ms is an eternity to a machine

This is incorrect for most of our CI test machines as well as most Azure's A1s and A4s. Unless you're willing to test on those, please keep the current timeout.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 9 above)
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Yeah, but the exception should be thrown on the first outstanding `SendAsync` that occurs - additional invocations only increase the chance one of them fails (which is still potentially important, yes).  Problem was, the original data size, plus the way the test was written, hid the fact that _no exceptions were being thrown_.  I tried increasing the element count, but that didn't help; increasing the size did.

File: src/System.Net.WebSockets.Client/tests/CancelTest.cs (hunk 12 above)
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/CancelTest.cs]: > try re-running the test

This is not an acceptable outcome.  Tests should not fail randomly.  Such failures lead to non-trivial and negative impact on developer productivity.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 9 above)
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: > no exceptions were being thrown

That's an issue as well, so thank you for noticing it and increasing the size.  But that doesn't mean we should also reduce the number of tasks.  Please revert that.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 11 above)
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Hmmm, well, you're right that they can complete (which looked to have happened on Linux), if they manage to send all the data in the buffer before the second `Send` is called.  This was why I increased the size of the data sent, although I was aware it might not be completely future-proof.

The way to fix it, though, would be to tell the echo server to not accept the whole thing, or delay the send somehow.  Need to look into it more, but my first guess would be having it read/parse by token, instead of the entire message, might do it.

Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Argh, except the server is still picking up and buffering the connection, even if the server doesn't actually process it yet.
IOW:

1. send: delay 5 seconds - completes instantly
2. send: some text - completes instantly
3. receive: some text - completes after 5 seconds

...whereas I need step 2 to complete after 5 seconds.  Reading through the server code, it looks like the actual pickup/process on their side should be happening after 5 seconds, at least from that layer's view.

@stephentoub - relying on the size of the message sent isn't likely to work long term either (as connections speeds will increase over time).  Increasing the number helps the chances of a conflict, but still isn't a guarantee; I'd prefer it if there were some way to tell the server to delay pickup (which would remove most of the possibility for race conditions), but that looks like it would require writing the echo server on a lower layer.  Thoughts?  I haven't investigated the whole stack, so it's possible the send task is returned as "completed" even if it hasn't left the box yet, but I'm figuring this is unlikely.

**Comment by stephentoub** [comment]
@Clockwork-Muse, what's the status of this PR?

**Comment by stephentoub** [comment]
> still waiting to get a reaction to possibly modifying the echo server to be able to ignore/delay incoming frames. I really don't see a good long term way to ensure the error is thrown otherwise.

You're welcome to open an issue for that specifically. In the meantime, let's just revert your related changes.

File: src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 13]
```diff
@@ -144,59 +132,34 @@ public class SendReceiveTest : ClientWebSocketTestBase
...
-                var cts = new CancellationTokenSource(TimeOutMilliseconds);
-
-                Task[] tasks = new Task[2];
-
-                await cws.SendAsync(
-                    WebSocketData.GetBufferFromText(".delay5sec"),
-                    WebSocketMessageType.Text,
-                    true,
-                    cts.Token);
+                var tasks = new Task[2];
```
Comment (Clockwork-Muse) [inline src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: @stephentoub - more information about how this is not a race condition on windows (currently).
The critical part is in [WinHttpWebSocket](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs#L395-L446), but what ends up happening is that when the second call to `ReceiveAsync` happens, the second call sets up the exception on the TCS returned by the _first_ call, then also returns that same task:
```
if (receiveOperationAlreadyPending)
{
    var exception = new InvalidOperationException(
    SR.Format(SR.net_Websockets_AlreadyOneOutstandingOperation, "ReceiveAsync"));
    _operation.TcsReceive.TrySetException(exception);

    Abort();
}

return _operation.TcsReceive.Task;
```
(Given the locking that occurs, this is guaranteed to happen even if submitted on two concurrently running threads, although which one gets it would be up in the air)

This is clearly different from the behavior on Linux/elsewhere, when going through [ManagedWebSocket](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ManagedWebSocket.cs#L288-L309) it returns a separate task based on a thrown exception, leaving the first task to complete "on it's own" (shutdown gracefully, IIRC, since the connection ends up closed):
```
try
{
    WebSocketValidate.ThrowIfInvalidState(_state, _disposed, s_validReceiveStates);

    Debug.Assert(!Monitor.IsEntered(StateUpdateLock), $"{nameof(StateUpdateLock)} must never be held when acquiring {nameof(ReceiveAsyncLock)}");
    lock (ReceiveAsyncLock) // synchronize with receives in CloseAsync
    {
        ThrowIfOperationInProgress(_lastReceiveAsync);
        Task<WebSocketReceiveResult> t = ReceiveAsyncPrivate(buffer, cancellationToken);
        _lastReceiveAsync = t;
        return t;
        }
}
catch (Exception exc)
{
    return Task.FromException<WebSocketReceiveResult>(exc);
}
```
... soooo now what?  Even without knowing what's going on inside, I don't feel like I would describe this as a race condition, looking at it externally (Assuming that there's no chance for the call to resolve first).

Is this something we should special case out based on implementation used, or should one of the versions be changed to match the other?

**Comment by karelz** [comment]
@Clockwork-Muse any update on this PR?

**Comment by Clockwork-Muse** [comment]
@CIPop - sorry, I've updated with what I've got, and updated the name like you wanted.

**Comment by Clockwork-Muse** [comment]
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please

**Comment by stephentoub** [comment]
> Note that I'm still expecting the multiple receive to fail, since there's a major difference in the implementation. I don't know what the current plan for that is, though.

If there's a significant observable behavioral difference between the implementations, then please open an issue for that and it can be investigated separately.  In the meantime, please revert the relevant test changes. Thanks.

**Comment by stephentoub** [comment]
LGTM.  Thanks for iterating on this.
